Evision.Mat.update_roi crashing BEAM

Hi! Love this project, and hoping I can start to contribute sometime soon :) 

I am having some trouble using `Evision.Mat.update_roi`, and it results in crashing the BEAM (or at least my livebook runtime). I haven't figured out the correct parameters to pass in order to 'paste' one image on top of another. Similar numpy code would look like the following:

```python
def draw_image(image, image_to_draw, x, y):
    h, w = image_to_draw.shape[:2]
    image[y:y + h, x:x + w, :] = image_to_draw
    return image
```

Now, when I try to do the same with Evision, my implementation is below:

```elixir
  def draw_image(image, image_to_draw, x, y) do
    {h, w, _} = image_to_draw.shape
    roi = [{y, y+h}, {x, x+w}, {0, 2}]
    Evision.Mat.update_roi(image, roi, image_to_draw)
  end
```

This results in the Livebook runtime crashing and the terminal printing the following error: 

`libc++abi: terminating with uncaught exception of type cv::Exception: OpenCV(4.6.0) /Users/runner/work/evision/evision/3rd_party/opencv/opencv-4.6.0/modules/core/src/matrix.cpp:879: error: (-215:Assertion failed) (int)ranges.size() == d in function 'Mat`

It looks like I'm doing something wrong with my indexing? Or just the general format of the `roi` parameter.

I think the docs could use some clarification or examples on what the correct `roi` parameter looks like. Example: are the `{start_index, end_index}` of the ROI inclusive? (I'd be happy to contribute after getting some clarity on these points!). And also it would of course probably be good to catch the error as a `{:error, "invalid ROI"}` tuple :) 

Let me know what you think.

Thanks and happy new year!

Gus

EDIT:

It does not crash when I use the following implementation (which removes the third channel from the ROI), but it also doesn't update the ROI as I would expect:

```elixir
  def draw_image(image, image_to_draw, x, y) do
    {h, w, _} = image_to_draw.shape
    roi = [{y, y+h}, {x, x+w}]
    Evision.Mat.update_roi(image, roi, image_to_draw)
  end
```

--- a/.githooks/pre-commit
+++ b/.githooks/pre-commit
@@ -6,7 +6,7 @@
 
 codespell="$(which codespell)"
 if [ -x "${codespell}" ]; then
-  "${codespell}" -S _build,deps,./lib/generated,3rd_party,.elixir_ls,erl_crash.dump,doc,test,.git,.cache,cmake_opencv_*,cmake_evision,cover,./c_src/evision_generated_funcs.h,.githook -L usign
+  "${codespell}" -S _build,deps,./lib/generated,3rd_party,.elixir_ls,erl_crash.dump,doc,test,.git,.cache,cmake_opencv_*,cmake_evision,cover,./c_src/evision_generated_funcs.h,./c_src/evision_generated_types_content.h,.githook -L usign
   if [ $? -eq 0 ]; then
     exit 0
   else

--- a/CHANGELOG.md
+++ b/CHANGELOG.md
@@ -1,11 +1,19 @@
 # Changelog
 
+## v0.1.26 (2023-01-03)
+[Browse the Repository](https://github.com/cocoa-xu/evision/tree/v0.1.26) | [Released Assets](https://github.com/cocoa-xu/evision/releases/tag/v0.1.26)
+
+### Fixed
+- [Evision.Mat] fixed `Evision.Mat.update_roi/3`.
+
 ## v0.1.25 (2022-12-18)
+[Browse the Repository](https://github.com/cocoa-xu/evision/tree/v0.1.25) | [Released Assets](https://github.com/cocoa-xu/evision/releases/tag/v0.1.25)
 
 ### Fixed
 - [smartcell] fixed a typo in SFace.
 
 ## v0.1.23/v0.1.24 (2022-12-17)
+[Browse the Repository](https://github.com/cocoa-xu/evision/tree/v0.1.24) | [Released Assets](https://github.com/cocoa-xu/evision/releases/tag/v0.1.24)
 
 ### Fixed
 - [smartcell] fixed charset loading when initialising FP16/INT8 CRNN models. [#144](https://github.com/cocoa-xu/evision/issues/144)

--- a/README.md
+++ b/README.md
@@ -206,7 +206,7 @@ Then you can add `evision` as dependency in your `mix.exs`.
 ```elixir
 def deps do
   [
-    {:evision, "~> 0.1.25"}
+    {:evision, "~> 0.1.26"}
   ]
 end
 ```
@@ -221,11 +221,11 @@ The following environment variables can be set based on your needs.
 (Note that precompiled binaries do not use FFmpeg. If you'd like to use FFmpeg, please compile from source (please see instructions in the next section) and set corresponding environment variables. We're considering this option at the moment.)
 
 #### Important notes
-It is recommended to use `:evision` from hex.pm. Currently "0.1.7" to "0.1.9", and "0.1.11" to "0.1.25" are available on hex.pm,
+It is recommended to use `:evision` from hex.pm.
 ```elixir
 def deps do
   [
-    {:evision, "~> 0.1.25"}
+    {:evision, "~> 0.1.26"}
   ]
 end
 ```
@@ -277,7 +277,7 @@ export EVISION_PREFER_PRECOMPILED=false
 For livebook users, 
 ```elixir
 Mix.install([
-  {:evision, "~> 0.1.25"}
+  {:evision, "~> 0.1.26"}
 ], system_env: [
   {"EVISION_PREFER_PRECOMPILED", "false"}
 ])

--- a/c_src/modules/evision_mat.h
+++ b/c_src/modules/evision_mat.h
@@ -222,6 +222,34 @@ static ERL_NIF_TERM evision_cv_mat_roi(ErlNifEnv *env, int argc, const ERL_NIF_T
     else return enif_make_badarg(env);
 }
 
+/// `matrix` and `patch` should be contiguous.
+void mat_update_roi(cv::Mat& matrix, cv::Mat& patch, 
+    std::vector<cv::Range>& ranges, size_t elem_size,
+    size_t step_size, size_t from_step_size,
+    int dim_index, size_t base_offset, size_t from_offset)
+{
+    if (dim_index >= matrix.size.dims()) return;
+
+    step_size /= matrix.size.p[dim_index];
+    from_step_size /= patch.size.p[dim_index];
+
+    auto& range = ranges[dim_index];
+
+    if (step_size == elem_size) {
+        uint8_t* data = matrix.data;
+        uint8_t* from = patch.data;
+        memcpy(
+            (uint64_t *)(((uint64_t)(uint64_t *)(data)) + base_offset + range.start * elem_size),
+            (uint64_t *)(((uint64_t)(uint64_t *)(from)) + from_offset),
+            elem_size * (range.end - range.start)
+        );
+    } else {
+        for (int pos = range.start, from_pos = 0; pos < range.end; pos++, from_pos++) {
+            mat_update_roi(matrix, patch, ranges, elem_size, step_size, from_step_size, dim_index + 1, base_offset + step_size * pos, from_offset + from_step_size * from_pos);
+        }
+    }
+}
+
 // @evision c: mat_update_roi,evision_cv_mat_update_roi,1
 // @evision nif: def mat_update_roi(_opts \\ []), do: :erlang.nif_error("Mat::update_roi() not loaded")
 static ERL_NIF_TERM evision_cv_mat_update_roi(ErlNifEnv *env, int argc, const ERL_NIF_TERM argv[]) {
@@ -242,40 +270,25 @@ static ERL_NIF_TERM evision_cv_mat_update_roi(ErlNifEnv *env, int argc, const ER
         evision_to_safe(env, evision_get_kw(env, erl_terms, "with_mat"), with_mat, ArgInfo("with_mat", 0)) &&
         evision_to_safe(env, evision_get_kw(env, erl_terms, "ranges"), ranges, ArgInfo("ranges", 0)))
     {
-        Mat roi = mat(ranges);
-        if (roi.channels() != with_mat.channels()) {
-            // `with_mat` could be an Nx.tensor
-            // hence the last dim should actually be the channel
-            int with_mat_dims = with_mat.size.dims();
-            int roi_dims = roi.size.dims();
-
-            // basically last_dim_as_channel
-            bool last_dim_as_channel = true;
-            if (with_mat_dims == roi_dims + 1 && with_mat.size.p[with_mat_dims - 1] == roi.channels()) {
-                for (int i = 0; i < roi_dims; i++) {
-                    if (with_mat.size.p[i] != roi.size.p[i]) {
-                        last_dim_as_channel = false;
-                        break;
-                    }
-                }
-            }
-
-            if (last_dim_as_channel) {
-                // we can still fix this
-                std::vector<int> new_shape(with_mat_dims - 1);
-                for (size_t i = 0; i < (size_t)(with_mat_dims - 1); i++) {
-                    new_shape[i] = with_mat.size.p[i];
-                }
-                int type = CV_MAKETYPE(with_mat.type(), with_mat.size.p[with_mat_dims - 1]);
-
-                // zero-copy constructor
-                Mat copy_from = Mat(with_mat_dims - 1, new_shape.data(), type, with_mat.data);
-                ERRWRAP2(copy_from.copyTo(roi), env, error_flag, error_term);
-            } else {
-                error_term = evision::nif::error(env, "Cannot update roi: shape mismatch");
-            }
+        if (mat.type() != with_mat.type()) {
+            error_flag = true;
+            error_term = evision::nif::error(env, "Cannot update roi: type mismatch");
+        } else if (mat.dims != with_mat.dims) {
+            error_flag = true;
+            error_term = evision::nif::error(env, "Cannot update roi: dimension mismatch");
         } else {
-            ERRWRAP2(with_mat.copyTo(roi), env, error_flag, error_term);
+            if (!mat.isContinuous()) mat = mat.clone();
+            if (!with_mat.isContinuous()) with_mat = with_mat.clone();
+            size_t step_size = mat.elemSize();
+            for (int i = 0; i < mat.size.dims(); i++) {
+                step_size *= mat.size.p[i];
+            }
+            size_t from_step_size = mat.elemSize();
+            for (int i = 0; i < with_mat.size.dims(); i++) {
+                from_step_size *= with_mat.size.p[i];
+            }
+            size_t elem_size = mat.elemSize();
+            mat_update_roi(mat, with_mat, ranges, elem_size, step_size, from_step_size, 0, 0, 0);
         }
 
         if (!error_flag) {

--- a/examples/cifar10.livemd
+++ b/examples/cifar10.livemd
@@ -11,7 +11,7 @@
 # System.put_env("EVISION_PREFER_PRECOMPILED", "false")
 
 Mix.install([
-  {:evision, "~> 0.1.25"},
+  {:evision, "~> 0.1.26"},
   {:req, "~> 0.3"},
   {:torchx, "~> 0.3"},
   {:nx, "~> 0.3", override: true},

--- a/examples/densenet121_benchmark.livemd
+++ b/examples/densenet121_benchmark.livemd
@@ -11,7 +11,7 @@
 # System.put_env("EVISION_PREFER_PRECOMPILED", "false")
 
 Mix.install([
-  {:evision, "~> 0.1.25"},
+  {:evision, "~> 0.1.26"},
   {:kino, "~> 0.7"},
   {:req, "~> 0.3"}
 ])

--- a/examples/dnn-detection-model.livemd
+++ b/examples/dnn-detection-model.livemd
@@ -11,7 +11,7 @@
 # System.put_env("EVISION_PREFER_PRECOMPILED", "false")
 
 Mix.install([
-  {:evision, "~> 0.1.25"},
+  {:evision, "~> 0.1.26"},
   {:req, "~> 0.3"},
   {:kino, "~> 0.7"}
 ])

--- a/examples/dnn-googlenet.livemd
+++ b/examples/dnn-googlenet.livemd
@@ -11,7 +11,7 @@
 # System.put_env("EVISION_PREFER_PRECOMPILED", "false")
 
 Mix.install([
-  {:evision, "~> 0.1.25"},
+  {:evision, "~> 0.1.26"},
   {:kino, "~> 0.7"},
   {:req, "~> 0.3"}
 ])

--- a/examples/find_and_draw_contours.livemd
+++ b/examples/find_and_draw_contours.livemd
@@ -9,7 +9,7 @@
 # System.put_env("EVISION_PREFER_PRECOMPILED", "false")
 
 Mix.install([
-  {:evision, "~> 0.1.25"},
+  {:evision, "~> 0.1.26"},
   {:kino, "~> 0.7"},
   {:req, "~> 0.3"}
 ])

--- a/examples/getting_started.livemd
+++ b/examples/getting_started.livemd
@@ -11,7 +11,7 @@
 # System.put_env("EVISION_PREFER_PRECOMPILED", "false")
 
 Mix.install([
-  {:evision, "~> 0.1.25"},
+  {:evision, "~> 0.1.26"},
   {:kino, "~> 0.7"},
   {:req, "~> 0.3"}
 ])

--- a/examples/ml-decision_tree_and_random_forest.livemd
+++ b/examples/ml-decision_tree_and_random_forest.livemd
@@ -11,7 +11,7 @@
 # System.put_env("EVISION_PREFER_PRECOMPILED", "false")
 
 Mix.install([
-  {:evision, "~> 0.1.25"},
+  {:evision, "~> 0.1.26"},
   {:kino, "~> 0.7"},
   {:scidata, "~> 0.1"},
   {:nx, "~> 0.4", override: true},

--- a/examples/ml-svm.livemd
+++ b/examples/ml-svm.livemd
@@ -11,7 +11,7 @@
 # System.put_env("EVISION_PREFER_PRECOMPILED", "false")
 
 Mix.install([
-  {:evision, "~> 0.1.25"},
+  {:evision, "~> 0.1.26"},
   {:kino, "~> 0.7"},
   {:req, "~> 0.3"}
 ])

--- a/examples/ocl-default_opencl_device_info.livemd
+++ b/examples/ocl-default_opencl_device_info.livemd
@@ -4,7 +4,7 @@
 
 ```elixir
 Mix.install([
-  {:evision, "~> 0.1.25"}
+  {:evision, "~> 0.1.26"}
 ])
 ```
 

--- a/examples/pca.livemd
+++ b/examples/pca.livemd
@@ -11,7 +11,7 @@
 # System.put_env("EVISION_PREFER_PRECOMPILED", "false")
 
 Mix.install([
-  {:evision, "~> 0.1.25"},
+  {:evision, "~> 0.1.26"},
   {:kino, "~> 0.7"},
   {:req, "~> 0.3"}
 ])

--- a/examples/photo-hdr.livemd
+++ b/examples/photo-hdr.livemd
@@ -11,7 +11,7 @@
 # System.put_env("EVISION_PREFER_PRECOMPILED", "false")
 
 Mix.install([
-  {:evision, "~> 0.1.25"},
+  {:evision, "~> 0.1.26"},
   {:kino, "~> 0.7"},
   {:req, "~> 0.3"}
 ])

--- a/examples/qrcode.livemd
+++ b/examples/qrcode.livemd
@@ -11,7 +11,7 @@
 # System.put_env("EVISION_PREFER_PRECOMPILED", "false")
 
 Mix.install([
-  {:evision, "~> 0.1.25"},
+  {:evision, "~> 0.1.26"},
   {:kino, "~> 0.7"},
   {:req, "~> 0.3"}
 ])

--- a/examples/stitching.livemd
+++ b/examples/stitching.livemd
@@ -11,7 +11,7 @@
 # System.put_env("EVISION_PREFER_PRECOMPILED", "false")
 
 Mix.install([
-  {:evision, "~> 0.1.25"},
+  {:evision, "~> 0.1.26"},
   {:kino, "~> 0.7"},
   {:req, "~> 0.3"}
 ])

--- a/examples/sudoku.livemd
+++ b/examples/sudoku.livemd
@@ -4,7 +4,7 @@
 
 ```elixir
 Mix.install([
-  {:evision, "~> 0.1.25"},
+  {:evision, "~> 0.1.26"},
   {:kino, "~> 0.7"},
   {:req, "~> 0.3"},
   {:torchx, "~> 0.4"},

--- a/examples/warp_perspective.livemd
+++ b/examples/warp_perspective.livemd
@@ -11,7 +11,7 @@
 # System.put_env("EVISION_PREFER_PRECOMPILED", "false")
 
 Mix.install([
-  {:evision, "~> 0.1.25"},
+  {:evision, "~> 0.1.26"},
   {:kino, "~> 0.7"},
   {:req, "~> 0.3"}
 ])

--- a/examples/warp_polar.livemd
+++ b/examples/warp_polar.livemd
@@ -9,7 +9,7 @@
 # System.put_env("EVISION_PREFER_PRECOMPILED", "false")
 
 Mix.install([
-  {:evision, "~> 0.1.25"},
+  {:evision, "~> 0.1.26"},
   {:kino, "~> 0.7"},
   {:req, "~> 0.3"}
 ])

--- a/lib/evision_mat.ex
+++ b/lib/evision_mat.ex
@@ -326,12 +326,44 @@ defmodule Evision.Mat do
   @spec update_roi(maybe_mat_in(), [{integer(), integer()} | Range.t() | :all], maybe_mat_in()) ::
           maybe_mat_out()
   def update_roi(mat, ranges, with_mat) do
-    mat = __from_struct__(mat)
-    ranges = __standardise_range_list__(ranges, true)
-    with_mat = __from_struct__(with_mat)
+    {mat, bring_back} =
+      if mat.dims != tuple_size(mat.shape) do
+        {Evision.Mat.channel_as_last_dim(mat), true}
+      else
+        {mat, false}
+      end
+    with_mat =
+      if with_mat.dims != tuple_size(with_mat.shape) do
+        Evision.Mat.channel_as_last_dim(with_mat)
+      else
+        with_mat
+      end
 
-    :evision_nif.mat_update_roi(mat: mat, ranges: ranges, with_mat: with_mat)
-    |> Evision.Internal.Structurise.to_struct()
+    ranges = __standardise_range_list__(ranges, true)
+    ranges =
+      if tuple_size(mat.shape) > Enum.count(ranges) do
+        extend =
+          for i <- Enum.count(ranges)..tuple_size(mat.shape)-1, reduce: [] do
+            acc ->
+              [{0, elem(mat.shape, i)} | acc]
+          end
+        ranges ++ Enum.reverse(extend)
+      else
+        ranges
+      end
+
+    with_mat = __from_struct__(with_mat)
+    mat = __from_struct__(mat)
+
+    res = Evision.Internal.Structurise.to_struct(
+      :evision_nif.mat_update_roi(mat: mat, ranges: ranges, with_mat: with_mat)
+    )
+
+    if bring_back do
+      Evision.Mat.last_dim_as_channel(res)
+    else
+      res
+    end
   end
 
   @doc """

--- a/mix.exs
+++ b/mix.exs
@@ -2,7 +2,7 @@ defmodule Evision.MixProject.Metadata do
   @moduledoc false
 
   def app, do: :evision
-  def version, do: "0.1.25"
+  def version, do: "0.1.26"
   def github_url, do: "https://github.com/cocoa-xu/evision"
   def opencv_version, do: "4.7.0"
   # only means compatible. need to write more tests
@@ -884,7 +884,7 @@ defmodule Evision.MixProject do
 
   defp generate_cmake_options() do
     mc = module_configuration()
-    enable_opencv_contrib = true
+    enable_opencv_contrib = false
     all_modules = Enum.map(mc.opencv, fn {m, _} -> m end) ++ Enum.map(mc.opencv_contrib, fn {m, _} -> m end)
     enabled_modules = Enum.filter(mc.opencv, fn {_, e} -> e end)
       ++ (if enable_opencv_contrib do Enum.filter(mc.opencv_contrib, fn {_, e} -> e end) else [] end)

--- a/src/evision.app.src
+++ b/src/evision.app.src
@@ -1,6 +1,6 @@
 {application, evision,
  [{description, "OpenCV-Erlang/Elixir binding."},
-  {vsn, "0.1.25"},
+  {vsn, "0.1.26"},
   {registered, []},
   {applications,
    [kernel,

--- a/test/evision_mat_test.exs
+++ b/test/evision_mat_test.exs
@@ -86,4 +86,43 @@ defmodule Evision.Mat.Test do
     bgr = Evision.cvtColor(transposed, Evision.cv_COLOR_RGB2BGR())
     assert {441, 297, 3} = Evision.Mat.shape(bgr)
   end
+
+  @tag :nx
+  test "update_roi" do
+    image_tensor = Evision.Mat.literal([[[2,2,2]], [[2,3,4]], [[4,5,6]]], :u8)
+    image_2d = Evision.Mat.last_dim_as_channel(image_tensor)
+    Evision.Mat.to_nx(image_2d)
+
+    patch_tensor = Evision.Mat.literal([[[7,8]], [[9,10]]], :u8)
+    patch_2d = Evision.Mat.last_dim_as_channel(patch_tensor)
+    Evision.Mat.to_nx(patch_2d)
+
+    flatlist = [2, 7, 8, 2, 9, 10, 4, 5, 6]
+    roi = [{0, 2}, {0, 1}, {1, 3}]
+    assert flatlist == Nx.to_flat_list(Evision.Mat.to_nx(Evision.Mat.update_roi(image_2d, roi, patch_2d)))
+    assert flatlist == Nx.to_flat_list(Evision.Mat.to_nx(Evision.Mat.update_roi(image_2d, roi, patch_tensor)))
+    assert flatlist == Nx.to_flat_list(Evision.Mat.to_nx(Evision.Mat.update_roi(image_tensor, roi, patch_2d)))
+    assert flatlist == Nx.to_flat_list(Evision.Mat.to_nx(Evision.Mat.update_roi(image_tensor, roi, patch_tensor)))
+
+    flatlist = [2, 2, 2, 2, 7, 8, 4, 9, 10]
+    roi = [{1, 3}, {0, 1}, {1, 3}]
+    assert flatlist == Nx.to_flat_list(Evision.Mat.to_nx(Evision.Mat.update_roi(image_2d, roi, patch_2d)))
+    assert flatlist == Nx.to_flat_list(Evision.Mat.to_nx(Evision.Mat.update_roi(image_2d, roi, patch_tensor)))
+    assert flatlist == Nx.to_flat_list(Evision.Mat.to_nx(Evision.Mat.update_roi(image_tensor, roi, patch_2d)))
+    assert flatlist == Nx.to_flat_list(Evision.Mat.to_nx(Evision.Mat.update_roi(image_tensor, roi, patch_tensor)))
+
+    flatlist = [2, 2, 2, 7, 8, 4, 9, 10, 6]
+    roi = [{1, 3}, {0, 1}, {0, 2}]
+    assert flatlist == Nx.to_flat_list(Evision.Mat.to_nx(Evision.Mat.update_roi(image_2d, roi, patch_2d)))
+    assert flatlist == Nx.to_flat_list(Evision.Mat.to_nx(Evision.Mat.update_roi(image_2d, roi, patch_tensor)))
+    assert flatlist == Nx.to_flat_list(Evision.Mat.to_nx(Evision.Mat.update_roi(image_tensor, roi, patch_2d)))
+    assert flatlist == Nx.to_flat_list(Evision.Mat.to_nx(Evision.Mat.update_roi(image_tensor, roi, patch_tensor)))
+
+    flatlist = [7, 8, 2, 9, 10, 4, 4, 5, 6]
+    roi = [{0, 2}, {0, 1}, {0, 2}]
+    assert flatlist == Nx.to_flat_list(Evision.Mat.to_nx(Evision.Mat.update_roi(image_2d, roi, patch_2d)))
+    assert flatlist == Nx.to_flat_list(Evision.Mat.to_nx(Evision.Mat.update_roi(image_2d, roi, patch_tensor)))
+    assert flatlist == Nx.to_flat_list(Evision.Mat.to_nx(Evision.Mat.update_roi(image_tensor, roi, patch_2d)))
+    assert flatlist == Nx.to_flat_list(Evision.Mat.to_nx(Evision.Mat.update_roi(image_tensor, roi, patch_tensor)))
+  end
 end

